### PR TITLE
Fully functional multitenant server

### LINKED ISSUE #65: Separate API packages
The 3 categories under API in the readme
- Helm Chart Repository
- Chart Manipulation
- Server Info

should each have their own subpackage/file

### LINKED ISSUE #78: Enable CRUD API (Chart Manipulation routes) in multitenant server
We could likely get away with using the same routes:
- `POST /api/charts` - upload a new chart version
- `POST /api/prov` - upload a new provenance file
- `DELETE /api/charts/<name>/<version>` - delete a chart version
- `GET /api/charts` - list all charts
- `GET /api/charts/<name>` - list all versions of a chart
- `GET /api/charts/<name>/<version>` - describe a chart version

We just need a way to pass in the name of the repo. Should we do so in the headers?
```
ChartMuseum-Repo: org1/repo1
```
or eliminate the `--data-binary` option and make it a form field (@davidovich)?
```
curl -F "repo=org1/repo1" -F "chart=@mychart-0.1.0.tgz" http://localhost:8080/api/charts
```

?
cc @alexei-led @itaysk

### LINKED ISSUE #82: Single tenant server, --multitenant flag should be removed
Once multitenant server is ready for prime time, it should be made the only server package. Singletenancy is supported in the multitenancy via `--depth=0`.

### PR DESCRIPTION
**Pull request by jdolitsky**
Removed the old singletenant codepath, making multitenant server the default and only option. All features supported in singletenant are now supported in multitenant. The CRUD API is now fully functional.

depth=0 by default, so it will continue to function by default as a "singletenant" server.

Lots of cleanup and reorganization as well, especially related to route handling.

Test coverage report:
```
github.com/kubernetes-helm/chartmuseum/cmd/chartmuseum/main.go (97.6%)
github.com/kubernetes-helm/chartmuseum/pkg/chartmuseum/logger/logger.go (97.0%)
github.com/kubernetes-helm/chartmuseum/pkg/chartmuseum/router/authorization.go (100.0%)
github.com/kubernetes-helm/chartmuseum/pkg/chartmuseum/router/match.go (100.0%)
github.com/kubernetes-helm/chartmuseum/pkg/chartmuseum/router/middleware.go (100.0%)
github.com/kubernetes-helm/chartmuseum/pkg/chartmuseum/router/router.go (88.9%)
github.com/kubernetes-helm/chartmuseum/pkg/chartmuseum/server.go (83.3%)
github.com/kubernetes-helm/chartmuseum/pkg/chartmuseum/server/multitenant/api.go (96.2%)
github.com/kubernetes-helm/chartmuseum/pkg/chartmuseum/server/multitenant/cache.go (91.9%)
github.com/kubernetes-helm/chartmuseum/pkg/chartmuseum/server/multitenant/handlers.go (92.6%)
github.com/kubernetes-helm/chartmuseum/pkg/chartmuseum/server/multitenant/index.go (80.0%)
github.com/kubernetes-helm/chartmuseum/pkg/chartmuseum/server/multitenant/routes.go (100.0%)
github.com/kubernetes-helm/chartmuseum/pkg/chartmuseum/server/multitenant/server.go (91.7%)
github.com/kubernetes-helm/chartmuseum/pkg/chartmuseum/server/multitenant/storage.go (100.0%)
github.com/kubernetes-helm/chartmuseum/pkg/repo/chart.go (97.4%)
github.com/kubernetes-helm/chartmuseum/pkg/repo/index.go (95.3%)
github.com/kubernetes-helm/chartmuseum/pkg/repo/metrics.go (100.0%)
github.com/kubernetes-helm/chartmuseum/pkg/repo/provenance.go (100.0%)
github.com/kubernetes-helm/chartmuseum/pkg/storage/alibaba.go (90.5%)
github.com/kubernetes-helm/chartmuseum/pkg/storage/amazon.go (97.6%)
github.com/kubernetes-helm/chartmuseum/pkg/storage/google.go (91.7%)
github.com/kubernetes-helm/chartmuseum/pkg/storage/local.go (94.1%)
github.com/kubernetes-helm/chartmuseum/pkg/storage/microsoft.go (80.6%)
github.com/kubernetes-helm/chartmuseum/pkg/storage/storage.go (100.0%)
```

Fixes #65 
Fixes #78 
Fixes #82

### CODE REVIEW

File: pkg/chartmuseum/router/authorization.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	if router.BasicAuthHeader != "" {
+		if router.AnonymousGet && request.Method == "GET" {
+			authorized = true
+		} else if request.Header.Get("Authorization") == router.BasicAuthHeader {
+			authorized = true
+		} else {
+			responseHeaders["WWW-Authenticate"] = "Basic realm=\"ChartMuseum\""
+		}
+	} else {
+		authorized = true
```
Comment (liamawhite) [inline pkg/chartmuseum/router/authorization.go]: if no basic auth header, just allow? Am I missing something?

File: pkg/chartmuseum/router/match.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+--depth=2:
+	Path: "/myorg/myrepo/index.yaml"
+	Repo: "myorg/myrepo"
+--depth=3:
+	Path: "/myorg/myteam/myrepo/index.yaml"
+	Repo: "myorg/myteam/myrepo"
+*/
+func match(routes []*Route, method string, url string, contextPath string, depth int) (*Route, []gin.Param) {
+	var noRepoPathSplit []string
+	var repo string
```
Comment (liamawhite) [inline pkg/chartmuseum/router/match.go]: Nit: define some of these on the same line `var repo, repoPath, noRepoPath string`

File: pkg/chartmuseum/router/match_test.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+			}
+			val, exists = c.Get("index")
+			suite.True(exists)
+			suite.Equal(9, val)
+			suite.Equal([]gin.Param{{"name", "mychart"}, {"version", "0.1.0"}, {"repo", repo}}, params)
+		}
+	}
+}
+
+func TestMatchTestSuite(t *testing.T) {
```
Comment (liamawhite) [inline pkg/chartmuseum/router/match_test.go]: http://cs-guy.com/blog/2015/01/test-main/

Comment (liamawhite) [reply pkg/chartmuseum/router/match_test.go]: Test Main is more idiomatic but it's not a big deal ^^

File: pkg/chartmuseum/server/multitenant/routes.go [hunk 4]
```diff
@@ -4,21 +4,34 @@ import (
...
-	helmChartRepositoryRoutes := []cm_router.Route{
-		{"READ", "GET", "/:repo/index.yaml", s.getIndexFileRequestHandler},
-		{"READ", "GET", "/:repo/charts/:filename", s.getStorageObjectRequestHandler},
+	helmChartRepositoryRoutes := []*cm_router.Route{
+		{"GET", "/:repo/index.yaml", s.getIndexFileRequestHandler, cm_router.RepoPullAction},
+		{"GET", "/:repo/charts/:filename", s.getStorageObjectRequestHandler, cm_router.RepoPullAction},
+	}
+
+	chartManipulationRoutes := []*cm_router.Route{
+		{"GET", "/api/:repo/charts", s.getAllChartsRequestHandler, cm_router.RepoPullAction},
```
Comment (liamawhite) [inline pkg/chartmuseum/server/multitenant/routes.go]: Maybe these constants should be RepoReadAction and RepoWriteAction?

File: pkg/chartmuseum/router/authorization.go (hunk 1 above)
Comment (jdolitsky) [reply pkg/chartmuseum/router/authorization.go]: router.BasicAuthHeader is only set if the user is trying to supply basic auth protection

File: pkg/chartmuseum/router/authorization.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+func generateBasicAuthHeader(username string, password string) string {
+	base := username + ":" + password
+	basicAuthHeader := "Basic " + base64.StdEncoding.EncodeToString([]byte(base))
+	return basicAuthHeader
+}
+
+func (router *Router) authorizeRequest(request *http.Request) (bool, map[string]string) {
+	var authorized bool
```
Comment (liamawhite) [inline pkg/chartmuseum/router/authorization.go]: could you make this explicit for readability (`authorized := false`)

File: pkg/chartmuseum/server/multitenant/routes.go (hunk 4 above)
Comment (jdolitsky) [reply pkg/chartmuseum/server/multitenant/routes.go]: I think "pull" and "push" are actually more appropriate here, if we are trying to closely match Dockers JWT format:
```
{
    "iss": "auth.docker.com",
    "sub": "jlhawn",
    "aud": "registry.docker.com",
    "exp": 1415387315,
    "nbf": 1415387015,
    "iat": 1415387015,
    "jti": "tYJCO1c6cnyy7kAn0c7rKPgbV1H1bFws",
    "access": [
        {
            "type": "repository",
            "name": "samalba/my-app",
            "actions": [
                "pull",
                "push"
            ]
        }
    ]
}
```
https://docs.docker.com/registry/spec/auth/jwt/

File: pkg/chartmuseum/router/match_test.go (hunk 3 above)
Comment (jdolitsky) [reply pkg/chartmuseum/router/match_test.go]: Ah. I am just following the convention in other test.go files. Lets update in near future for all.

**Pull request merged**